### PR TITLE
Handle validation of uxtheme.dll undocumented exported functions in ARM64 builds

### PR DESCRIPTION
**Pull request by hmartinez82**
This is required for https://github.com/eclipse-platform/eclipse.platform.swt/pull/1045 to continue.

For reference, here are the disassemblies:
`AllowDarkModeForWindow()`:
```
00007FFB16C32450 D503237F             pacibsp  
00007FFB16C32454 A9BE53F3             stp         x19,x20,[sp,#-0x20]!  
00007FFB16C32458 F9000BF5             str         x21,[sp,#0x10]  
00007FFB16C3245C A9BF7BFD             stp         fp,lr,[sp,#-0x10]!  
00007FFB16C32460 910003FD             mov         fp,sp  
00007FFB16C32464 53001C33             uxtb        w19,w1  
00007FFB16C32468 D29523C1             mov         x1,#0xA91E  
00007FFB16C3246C AA0003F5             mov         x21,x0  
00007FFB16C32470 9400F226             bl          #GetPropW (07FFB16C6ED08h)
...
```
`AllowDarkModeForWindowWithTelemetryId():`
```
00007FFB16C33210 D503237F             pacibsp  
00007FFB16C33214 A9BE53F3             stp         x19,x20,[sp,#-0x20]!  
00007FFB16C33218 F9000BF5             str         x21,[sp,#0x10]  
00007FFB16C3321C A9BF7BFD             stp         fp,lr,[sp,#-0x10]!  
00007FFB16C33220 910003FD             mov         fp,sp  
00007FFB16C33224 53001C34             uxtb        w20,w1  
00007FFB16C33228 D29523C1             mov         x1,#0xA91E  
00007FFB16C3322C AA0003F5             mov         x21,x0  
00007FFB16C33230 9400EEB6             bl          #GetPropW (07FFB16C6ED08h)
...
```
`SetPreferredAppMode()`:
```
00007FFB16C33E10 D503237F             pacibsp  
00007FFB16C33E14 F81F0FF3             str         x19,[sp,#-0x10]!  
00007FFB16C33E18 A9BF7BFD             stp         fp,lr,[sp,#-0x10]!  
00007FFB16C33E1C 910003FD             mov         fp,sp  
00007FFB16C33E20 D00007E8             adrp        x8,wil::details::g_enabledStateManager+40h (07FFB16D31000h)  
00007FFB16C33E24 B94BD913             ldr         w19,[x8,#0xBD8]  
00007FFB16C33E28 2A0003E1             mov         w1,w0  
00007FFB16C33E2C 912F6100             add         x0,x8,#0xBD8  
00007FFB16C33E30 97FF7910             bl          _InterlockedExchange (07FFB16C12270h)  
00007FFB16C33E34 2A1303E0             mov         w0,w19  
00007FFB16C33E38 A8C17BFD             ldp         fp,lr,[sp],#0x10  
00007FFB16C33E3C F84107F3             ldr         x19,[sp],#0x10  
00007FFB16C33E40 D50323FF             autibsp  
00007FFB16C33E44 D65F03C0             ret  
00007FFB16C33E48 B0000090             adrp        xip0,GetBufferedPaintBitsEx+7320h (07FFB16C44000h)
...
```

### CODE REVIEW

**Comment by SyntevoAlex** [comment]
The code makes sense to me.
But I'm judging blindly without testing, let's wait until it can be smoke tested.

**Comment by hmartinez82** [comment]
> The code makes sense to me. But I'm judging blindly without testing, let's wait until it can be smoke tested.

**Comment by hmartinez82** [comment]
How do you want to smoke test? :) these changes don't build standalone. They need @chirontt changes to be built. I thought the idea was for this to be merged so he can merge his

**Comment by SyntevoAlex** [comment]
He will merge `return FALSE`, then your branch is rebased on top and tesred.

**Comment by chirontt** [comment]
@hmartinez82 I've temporarily merged these changes of `os_custom.c` file to my PR and attempted to 'build' it, but the MSVC compiler complains about `reinterpret_cast`:
```
[INFO]      [exec] Microsoft (R) C/C++ Optimizing Compiler Version 19.38.33135 for ARM64
[INFO]      [exec] Copyright (C) Microsoft Corporation.  All rights reserved.
[INFO]      [exec]
[INFO]      [exec]      cl -O1 /WX /W4 -DNDEBUG -DUNICODE -D_UNICODE /c   -DJNI64  -DSWT_VERSION=4964 -DSWT_REVISION=7 -DUSE_ASSEMBLER  /I"\apps\jdk-17.0.10+7_arm64\include" /I"\apps\jdk-17.0.10+7_arm64\include\win32" /I. os_custom.c
[INFO]      [exec] os_custom.c
[INFO]      [exec] os_custom.c(83): error C2065: 'reinterpret_cast': undeclared identifier
```

**Comment by hmartinez82** [comment]
omg, it's C code, not C++ . I'll fix it in a moment 😅

**Comment by hmartinez82** [comment]
@chirontt I've updated this PR with the new changes. They are the same, just using C casting instead of C++ ones.

**Comment by chirontt** [comment]
Yes, it now compiles properly and the SWT binaries are generated for Arm64:

image

@SyntevoAlex how do we smoke test it, i.e. maybe not about the `os_custom.c` file itself, but about the generated `*.dll` modules? Do we have some C/Java tests for them already in the repo?

**Comment by SyntevoAlex** [comment]
Snippet `Bug562043_DarkTableNoHover.java` should have dark scrollbars. But also set `org.eclipse.swt.internal.win32.useShellTitleColoring` and check titlebar,

**Comment by hmartinez82** [comment]
@SyntevoAlex I can't test this. It will have to be @chirontt. I'll close this since he's incorporating it in his pull request.

I can't test this because I don't knwk to build anything that uses maven, :( I don't know how

**Comment by chirontt** [comment]
@hmartinez82 you should wait for someone (@SyntevoAlex ?) to approve this PR and merge it to master, to claim credit for your excellent work. I don't plan to incorporate it into my PR. My plan is to remove this file from my PR.

**Comment by hmartinez82** [comment]
Ok, understood. It seems that they want you to merge yours first returning false then I'll rebsse this. 👍

Sent with Shortwave <https://www.shortwave.com?utm_medium=email&utm_content=signature&utm_source=aGVybmFuLmMubWFydGluZXpAZ21haWwuY29t>

**Comment by chirontt** [comment]
> Snippet `Bug562043_DarkTableNoHover.java` should have dark scrollbars (expand some Tree item for scrollbar to appear).
> 
> But also set `display.setData("org.eclipse.swt.internal.win32.useShellTitleColoring", true);` and check that Shell's titlebar is dark.

Yes, it works on my Arm64 box. I've built the `org.eclipse.swt.win32.win32.aarch64-3.125.0-SNAPSHOT.jar` which embeds the `swt*.dll` for Arm64, then run the `Bug562043_DarkTableNoHover.java` and it shows the dark scrollbars nicely, but with the lighted window title bar:

light-title

Then, as you suggested, I add the following statement to the Java code:
```java
		display.setData("org.eclipse.swt.internal.win32.useShellTitleColoring", true);
```
and I get a window with dark title bar:

dark-title

and out of curiosity, I comment out this existing statement:
```java
		display.setData("org.eclipse.swt.internal.win32.useDarkModeExplorerTheme", true);
```
then I get a window with both lighted title bar and scrollbars:

light-scrollbars

**Comment by chirontt** [comment]
Further more, I've merged this PR to my [Arm64 fork](https://github.com/chirontt/eclipse.platform.swt) of this repo and built the complete Eclipse SDK, via the [aggregator](https://github.com/chirontt/eclipse.platform.releng.aggregator), and here's how it looks in Arm64:

Eclipse SDK

For comparison, in previous build I used the `return FALSE;` statements in the `os_custom.c` for Arm64, and here's how it used to look:

Eclipse SDK light

I think my testing is enough to confirm the working of this PR for Arm64.

**Comment by SyntevoAlex** [comment]
Yes, it can be merged, then.
Can you please rebase the branch and/or edit commit message slightly to trigger a rebuild? Previous build failed for some weird reason.

**Comment by iloveeclipse** [comment]
> Yes, it can be merged, then.

Please note, we are between M3 and RC1 time frame. Any merge requires extra review & PM approval.
https://www.eclipse.org/lists/platform-releng-dev/msg39974.html

**Comment by SyntevoAlex** [comment]
Yes, I was checking mails right now... It says the feature freeze is on Feb 21 only?
Either way, I suggest to merge this PR, because it's under `#ifdef` and therefore code changes may only affect arm64 build, which is not currently present.

@iloveeclipse please clarify what to do.

**Comment by hmartinez82** [comment]
Rebased / squashed as requested

**Comment by iloveeclipse** [comment]
> please clarify what to do.

Please add your "approving" review on this PR. 
Looking on code, I agree it shouldn't harm so OK to merge for RC1.

**Review by SyntevoAlex — APPROVED** [review]
Looks good to me!

**Comment by SyntevoAlex** [comment]
The branch built successfully.
I added my approval.

**Comment by SyntevoAlex** [comment]
Everything looks good to me, I suggest to merge.

**Comment by SyntevoAlex** [comment]
@iloveeclipse so, should I merge? Or approval from PM is still needed? Who is PM?
Also, oculd you please clarify if freeze is already in effect or only since Feb 21?

**Comment by merks** [comment]
If everyone agrees it's very safe good to go, then it also has my +1 as a PMC member.

**Comment by SyntevoAlex** [comment]
OK, merged.
@hmartinez82 thanks for your contribution!
It was a pleasure to see someone who's not scared of disasm :)

**Comment by hmartinez82** [comment]
@SyntevoAlex If you want to give me a gift. I wouldn't mind a local build of SmartGit for Windows on ARM64. You can even give it to me without the Java Runtime. I'll use Microsoft's Java 17 build for ARM64 :)

**Comment by tmssngr** [comment]
@hmartinez82 At the moment Windows/ARM has low priority for us. Please vote for https://smartgit.userecho.com/communities/1/topics/1416-windows-arm-support

**Comment by chirontt** [comment]
@hmartinez82 Bad news: with recent Windows update to my Arm64 box, like this build:

image

these validation code no longer works, i.e. I get lighted scrollbars rather than dark ones. I guess Microsoft has updated/changed the `uxtheme.dll` file, so we may need your magical disassembly touch again :)

**Comment by hmartinez82** [comment]
Ok. I have the same version. I'll see what I can do.

**Comment by hmartinez82** [comment]
@chirontt I'll create the PR but you'll have to test it. I don't even know if the ordianal 135 is still `SetPreferredAppMode()`, but here is the disassembly:
```
00007FFA2A1D3E10  pacibsp  
00007FFA2A1D3E14  str         x19,[sp,#-0x10]!  
00007FFA2A1D3E18  stp         fp,lr,[sp,#-0x10]!  
00007FFA2A1D3E1C  mov         fp,sp  
00007FFA2A1D3E20  adrp        x8,00007FFA2A2D0000  
00007FFA2A1D3E24  ldr         w19,[x8,#0xBB8]  
00007FFA2A1D3E28  mov         w1,w0  
00007FFA2A1D3E2C  add         x0,x8,#0xBB8  
00007FFA2A1D3E30  bl          00007FFA2A1B2270  
00007FFA2A1D3E34  mov         w0,w19  
00007FFA2A1D3E38  ldp         fp,lr,[sp],#0x10  
00007FFA2A1D3E3C  ldr         x19,[sp],#0x10  
00007FFA2A1D3E40  autibsp  
00007FFA2A1D3E44  ret 
```

**Comment by hmartinez82** [comment]
@chirontt @SyntevoAlex  https://github.com/eclipse-platform/eclipse.platform.swt/pull/1172

**Pull request merged**
